galley/pkg/config/analysis/analyzers/gateway/gateway.go

```diff
@@ -0,0 +1,122 @@
+	"istio.io/istio/galley/pkg/config/processor/metadata"
+	"istio.io/istio/galley/pkg/config/resource"
+)
+
+// GatewayAnalyzer checks the gateways associated with each virtual service
+type Analyzer struct{}
+
+var _ analysis.Analyzer = &Analyzer{}
+
+// Metadata implements analysis.Analyzer
+func (*Analyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name: "gateway.GatewayAnalyzer",
+		Inputs: collection.Names{
+			metadata.IstioNetworkingV1Alpha3Gateways,
+			metadata.K8SCoreV1Pods,
+			metadata.K8SCoreV1Services,
+		},
+	}
+}
+
+// Analyze implements Analyzer
+func (s *Analyzer) Analyze(c analysis.Context) {
+	c.ForEach(metadata.IstioNetworkingV1Alpha3Gateways, func(r *resource.Entry) bool {
+		s.analyzeGateway(r, c)
+		return true
+	})
+}
+
+func (*Analyzer) analyzeGateway(r *resource.Entry, c analysis.Context) {
+
+	gw := r.Item.(*v1alpha3.Gateway)
+
+	// Typically there will be a single istio-ingressgateway service, which will select
+	// the same ingress gateway pod workload as the Gateway resource.  If there are multiple
+	// Kubernetes services, and they offer different TCP port combinations, this validator will
+	// not report a problem if any service exposes the Gateway port.
+	servicePorts := map[uint32]bool{}
+	gwSelectorMatches := 0
+
+	// For pods selected by gw.Selector, find Services that select them and remember those ports
+	gwSelector := k8s_labels.SelectorFromSet(gw.Selector)
+	c.ForEach(metadata.K8SCoreV1Pods, func(rPod *resource.Entry) bool {
+		pod := rPod.Item.(*v1.Pod)
+		podLabels := k8s_labels.Set(pod.ObjectMeta.Labels)
+		if gwSelector.Matches(podLabels) {
+			gwSelectorMatches++
+			c.ForEach(metadata.K8SCoreV1Services, func(rSvc *resource.Entry) bool {
+				nsSvc, _ := rSvc.Metadata.Name.InterpretAsNamespaceAndName()
+				if nsSvc != pod.ObjectMeta.Namespace {
+					return true // Services only select pods in their namespace
+				}
+
+				service := rSvc.Item.(*v1.ServiceSpec)
+				// TODO I want to match service.Namespace to pod.ObjectMeta.Namespace
+				svcSelector := k8s_labels.SelectorFromSet(service.Selector)
+				if svcSelector.Matches(podLabels) {
+					for _, port := range service.Ports {
+						if port.Protocol == "TCP" {
+							servicePorts[uint32(port.Port)] = true
+						}
+					}
+				}
+				return true
+			})
+		}
+		return true
+	})
+
+	if gwSelectorMatches == 0 {
+		// We found no service for the Gateway's workload selector.  If the Gateway does select
+		// the Istio system ingress gateway complain about a missing referenced resource.  (We
+		// don't want to complain about missing system resources, because a user may want to analyze
+		// only his own application files.)
+		if len(gw.Selector) != 1 || gw.Selector["istio"] != "ingressgateway" {
+			c.Report(metadata.IstioNetworkingV1Alpha3Gateways, msg.NewReferencedResourceNotFound(r, "selector", gwSelector.String()))
+			return
+		}
+		// The unreferenced Ingress is the System ingress, pretend we have found it.
```

Comment:
Is this defined anywhere else we could refer to? I'm worried that hardcoding a port list here will be brittle.

Reply:
The ports are only defined in YAML at e.g. _install/kubernetes/istio.yaml_.  Agreed that is brittle, but we have been offering the same ports for a long time.  It seems like overkill to dynamically create "_ingressports.gen.go_" from Helm charts and compile in the current ingress.

Reply:
I was afraid that would be the answer. :) I agree that trying to be dynamic against the helm charts would be overkill. 

Let's do extract this as a constant, though, no need to redefine the list every trip through the loop.